### PR TITLE
fix(appwrite): make `meta.permissions` overrideable #5767

### LINKED ISSUE #5767: [BUG] Appwrite `dataProvider` meta.permissions (`@refinedev/appwrite`)
### Describe the bug

Hello,
how to override these default permissions in appwrite dataProvider? I don't want public permissions, only the permissions i set in meta at creation/edition.
```ts
  const { saveButtonProps, formProps, form } = useForm<HttpError>({
    meta: {
      readPermissions: [Permission.read(Role.user(identity.$id))],
      writePermissions: [Permission.write(Role.user(identity.$id))],
    },
  });
  ```
  
  I was able to walk-around it by making custom `dataProviderWrapper` that overrides the original `create` method
  ```ts
    return {
    ...originalDataProvider,
    create: async ({ resource, variables, meta }) => {
      const permissions = [
        // Permission.read(Role.any()),
        // Permission.write(Role.any()),
        ...(meta?.readPermissions ?? [Permission.read(Role.any())]),
        ...(meta?.writePermissions ?? [Permission.write(Role.any())]),
      ];
```

### Steps To Reproduce

1. create new app
```
npm create refine-app@latest -- --example data-provider-appwrite
```
2. use the following snippet with any roles.
```ts
  const { saveButtonProps, formProps, form } = useForm<HttpError>({
    meta: {
      readPermissions: [Permission.read(Role.user(identity.$id))],
      writePermissions: [Permission.read(Role.user(identity.$id))],
    },
  });
  ```

### Expected behavior

expected behavior is to be able to complete override the `meta` params.

### Packages

```json
    "@refinedev/antd": "^5.37.5",
    "@refinedev/appwrite": "^6.4.7",
    "@refinedev/core": "^4.48.0",
    "@refinedev/devtools": "^1.1.35",
    "@refinedev/kbar": "^1.3.7",
    "@refinedev/nextjs-router": "^6.0.1",
    "@refinedev/react-hook-form": "^4.8.15",
```

### Additional Context

sample after override

```ts
export const customDataProvider = (
  appwriteClient: Appwrite,
  options: { databaseId: string } = { databaseId: 'default' }
): Required<DataProvider> => {
  const { databaseId } = options;
  const database = new Databases(appwriteClient);
  const originalDataProvider: Required<DataProvider> = dataProvider(
    appwriteClient,
    {
      databaseId,
    }
  );

  return {
    ...originalDataProvider,
    create: async ({ resource, variables, meta }) => {
      const permissions = [
        // Permission.read(Role.any()),
        // Permission.write(Role.any()),
        ...(meta?.readPermissions ?? [Permission.read(Role.any())]), // new permisions
        ...(meta?.writePermissions ?? [Permission.write(Role.any())]), // new permisions
      ];

      const { $id, ...restData } = await database.createDocument(
        databaseId,
        resource,
        meta?.documentId ?? ID.unique(),
        variables as unknown as object,
        permissions
      );

      return {
        data: {
          id: $id,
          ...restData,
        },
      } as any;
    },
}
```

### PR DESCRIPTION
**Pull request by a-essawy**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://refine.dev/docs/guides-concepts/contributing/#commit-convention

## Bugs / Features

- [x] Related issue(s) linked
- [ ] Tests for the changes have been added
- [ ] Docs have been added / updated
- [x] Changesets have been added https://refine.dev/docs/guides-concepts/contributing/#creating-a-changeset

## What is the current behavior?

Users couldn't override the default `readPermissions` /  `writePermissions` values.

## What is the new behavior?

Users are able to override the default permissions by passing `options.defaultReadPermissions` / `options.defaultWritePermissions` OR by passing `meta?.readPermissions` / `meta?.writePermissions`.

If user passed `options.defaultReadPermissions` AND `meta?.readPermissions`, then the priority will be as following:

`meta?.readPermissions` if not? then `options.defaultReadPermissions` if not? then `Role.any()`.

fixes #5767 

## Notes for reviewers

This should be compatible with old users who are not passing `meta?.readPermissions` and expecting to get `Role.any()` by default.

### CODE REVIEW

File: packages/appwrite/src/dataProvider.ts [hunk 1]
```diff
@@ -6,14 +6,31 @@ import {
...
+    defaultReadPermissions: [],
+    defaultWritePermissions: [],
+  },
 ): Required<DataProvider> => {
   const { databaseId } = options;
 
   const database = new Databases(appwriteClient);
 
+  const defaultReadPermissions = options.defaultReadPermissions ?? [
+    Permission.read(Role.any()),
```
Comment (BatuhanW) [inline packages/appwrite/src/dataProvider.ts:28]: Let's remove these fallback values. Let's use empty array by default, unless provided.

**Review by BatuhanW — COMMENTED** [review]
Hey @abdelrahman-essawy thanks for the quick PR! Added a minor comment, let's get rid of fallback any permissions completely.

**Comment by abdelrahman-essawy** [comment]
@BatuhanW I did remove the fallback value, but note that test `packages/appwrite/test/create/index.spec.ts` now fails.

it only passes with 
```ts
     defaultWritePermissions: [Permission.write(Role.any())],
     defaultReadPermissions: [Permission.read(Role.any())]
```

**Comment by BatuhanW** [comment]
> @BatuhanW I did remove the fallback value, but note that test `packages/appwrite/test/create/index.spec.ts` now fails.
> 
> it only passes with `defaultWritePermissions`, `defaultReadPermissions` passed in options AND not empty array.
> 
> ```ts
> describe("create", () => {
>   it("correct response with meta", async () => {
>     const { data } = await dataProvider(client, {
>       databaseId: "632455a0b8d017403ce9",
>       defaultWritePermissions: [Permission.write(Role.any())],
>       defaultReadPermissions: [Permission.read(Role.any())]
>     }).create({
>       resource: "632455a55dc72e1aa016",
>       variables: {
>         title: "Lorem",
>       },
>     });
>     
>     expect(data.title).toEqual("Lorem");
>     expect(data.id).toBeTruthy();
>   });
> });
> ```

Yeah, we need to update tests too. And also ideally re-record them with nock.

**Comment by BatuhanW** [comment]
Hey @abdelrahman-essawy, we've merged `releases/april` branch, so your PR is closed. Could you re-open your PR?

**Comment by BatuhanW** [comment]
Hey @abdelrahman-essawy we would like to release this one with May Release. Do you have time to re-open this PR so we can get it released?

**Comment by abdelrahman-essawy** [comment]
@BatuhanW sure thing, I will re-open it and update the tests this weekend.

**Pull request closed without merging**